### PR TITLE
plugins/telescope: refactor

### PR DESCRIPTION
**Pull request by GaetanLepage**
Refactoring of the `plugins.telescope` module:
- Switch to `mkNeovimPlugin`
- Better handling of extensions:
  Adopt a similar approach as in `neotest`: Add a helper function to create the submodule for each extension.
  This reduces the amount of boilerplate code.
- Fix wrong options
- Add tests

### CODE REVIEW

**Comment by traxys** [comment]
Please describe in more details in a message what is a "better handling of extensions" and why would it mandate a 1000+ lines refactor

**Comment by GaetanLepage** [comment]
> Please describe in more details in a message what is a "better handling of extensions" and why would it mandate a 1000+ lines refactor

Yes, indeed !
Actually the enormous diff is caused by `git` not handling the file renames properly and counted them as deletion-creation...

File: plugins/telescope/default.nix [hunk 1]
```diff
@@ -59,78 +69,86 @@ in {
...
+
+      extensions = mkOption {
+        type = with types; attrsOf anything;
+        default = {};
+        # NOTE: We hide this option from the documentation as users should use the top-level
+        # `extensions.*` options.
+        # They can still directly change this `settings.adapters` option.
+        # In this case, they are responsible for explicitly installing the manually added extensions.
+        visible = false;
+      };
```
Comment (MattSturgeon) [inline plugins/telescope/default.nix:129]: Could this be visible to mitigate issues like #1218?

> Perhaps `plugins.telescope.enabledExtensions` could be made public or an `plugins.telescope.extraExtensions` could be introduced to handle other unsupported telescope extensions?

Or is there a better solution?

Comment (traxys) [reply plugins/telescope/default.nix:129]: wouldn't `plugins.telescope.settings.extensions` work for that? It's the main adavantage of RFC 42 style plugins, you should _not_ be limited to what nixvim provides

**Review by traxys — APPROVED** [review]

File: plugins/telescope/extensions/media-files.nix [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+  config,
+  pkgs,
+  ...
+}:
+with lib;
+  (
+    import ./_helpers.nix {inherit lib helpers config pkgs;}
+  )
+  .mkExtension {
+    name = "media-files";
```
Comment (MattSturgeon) [inline plugins/telescope/extensions/media-files.nix:13]: I get `error: The option `plugins.telescope.extensions.media_files' does not exist.`

I think this should've had a rename deprecation added.

File: plugins/telescope/default.nix [hunk 3]
```diff
@@ -5,37 +5,47 @@
...
+          (
+            either
+            str
+            (submodule {
+              options = {
+                action = mkOption {
+                  type = types.str;
+                };
+                mode = helpers.keymaps.mkModeOption "n";
+                options = helpers.keymaps.mapConfigOptions;
```
Comment (MattSturgeon) [inline plugins/telescope/default.nix:38]: Not sure if it's possible for attrsOf submodule, but ideally this should've had a rename deprecation too:

`error: The option `plugins.telescope.keymaps."<leader>bb".desc' does not exist`

File: plugins/telescope/extensions/media-files.nix (hunk 2 above)
Comment (GaetanLepage) [reply plugins/telescope/extensions/media-files.nix:13]: Was it necessary to rename it to `media-files`... Not sure now.

File: plugins/telescope/default.nix (hunk 3 above)
Comment (GaetanLepage) [reply plugins/telescope/default.nix:38]: Yes, this is not really possible unfortunately.
The only alternative is to make a `mkRemovedOptionModule` for the whole `keymaps` option to force the user to re-define it manually.

**Pull request merged**